include/triton-shared/AnalysisStructured/PtrAnalysis.h

```diff
@@ -0,0 +1,209 @@
+//===----------------------------------------------------------------------===//
+//
+// Copyright (c) Microsoft Corporation, Meta Platforms.
+// Licensed under the MIT license.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef TRITON_ANALYSISSTRUCTURED_PTRANALYSIS_H
+#define TRITON_ANALYSISSTRUCTURED_PTRANALYSIS_H
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+
+#include "triton-shared/Dialect/TritonStructured/IR/TritonStructuredDialect.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+
+#include <set>
+
+namespace mlir {
+
+class OpBuilder;
+
+namespace triton {
+
+const extern std::string ptrAnalysisAttr;
+
+// Data structure used to decode pointer arithmetics. offsets, sizes, and
+// strides are in unit of elements in a linearly laid-out memory, which is the
+// same as pointer arithmetic operations in Triton language. scalar is a
+// shortcut used when the entire state describes a single scalar value. source
+// is the base pointer.
+class PtrSState {
+
+public:
+  SmallVector<OpFoldResult> offsets;
+  SmallVector<OpFoldResult> sizes;
+  SmallVector<OpFoldResult> strides;
+  SmallVector<OpFoldResult> modulos;
+
+  Value source;
+  Value scalar;
+
+  int32_t getRank() const;
+
+  bool isEmpty() const;
+
+  bool hasModulo() const;
+
+  bool dimHasModulo(uint32_t dim) const;
+
+  // Process addition of two PtrSStates.
+  LogicalResult addState(const PtrSState &lhsState, const PtrSState &rhsState,
+                         Operation *op, OpBuilder &builder);
+
+  // Process multiplication of two PtrSStates
+  LogicalResult mulState(const PtrSState &lhsState, const PtrSState &rhsState,
+                         Operation *op, OpBuilder &builder);
+
+  tts::MakeTensorPtrOp createTTSMakeTensorPtrOp(OpBuilder &builder,
+                                                Location loc);
+
+  static void swap(PtrSState &&a, PtrSState &&b);
+};
+
+struct PtrAnalysis {
+  using IndexMapSet = std::map<int, std::set<int>>;
+
+  IndexMapSet levelToBlockArgIndex;
+  int level = 0;
+
+  llvm::SmallDenseMap<Value, PtrSState> knownPtrs;
+
+  IRMapping map;
```

Comment:
nit: i think this map should be rename to something a bit more descriptive since it's only used for ptr handling. perhaps something like `ptrMap` would be better